Made changes to get more than one batch from Cache

shubhrojroy opened the pull request:
- Added getAll method to Cache class which also accepts a bufferSize
- Implemented getAll in CacheWithData to recursively fetch all contiguous batches from the cache that fit in the provided bufferSize and return it as a single CastleMessageBatch.
- Added method to CastleMessageBatch object to create single CastleMessageBatch from a Seq of batches.
- Finally changed FetchDataProcessor to call the getAll method instead of get larger batches from the cache when available. It passes the consumer's bufferSize to cache.getAll.
- Added test for the same.

denisgrenader commented on src/test/scala/com/box/castle/router/kafkadispatcher/cache/CacheTest.scala:
```diff
@@ -394,4 +394,38 @@ class CacheTest extends Specification with Mockito with MockBatchTools {
+      batch.sizeInBytes shouldEqual b1.sizeInBytes + b2.sizeInBytes + b3.sizeInBytes
+      batch.offset shouldEqual b1.offset
+      batch.nextOffset shouldEqual b3.nextOffset
+    }
```
Add more tests.  I am going to use 500MB and 16MB max buffer size with 1MB sized messages for example, but scale properly to tests.

1. Make sure we respect the max buffer size.  That is to say if the cache is holding 500MB of batches, and there are 500 batches of 1MB in there, with max buffer size of 16MB, the cache should only return 16 batches worth of data.
   1. Sub variant - make sure this works properly at the beginning offset
   1. Sub variant - make sure this works properly if it's returning the last 16MB worth of messages and it's exactly flush with the end, i.e. we get exactly the last 16 messages
   1. Sub variant - make sure this works properly if it's returning the last 16MB of messages but the offset is 1 off, so we only get the last 15 messages
   1. Sub variant - make sure if the max batch size is 16MB, and the offset is 1 from the end, then we return only 1 batch back
   1. Sub variant - this should work properly in the middle of the 500mb, that is to say with an offset half way into the cache

shubhrojroy replied:
Done

denisgrenader commented on src/main/scala/com/box/castle/batch/CastleMessageBatch.scala:
```diff
@@ -90,6 +90,19 @@ object CastleMessageBatch {
+    * @param messageBatches
+    * @return
+    */
+  def apply(messageBatches: Vector[CastleMessageBatch]): CastleMessageBatch = {
```
this needs a runtime check to make sure that all the offsets in the vector are contiguous, or there will be a chance to miss some messages

shubhrojroy replied:
Done

denisgrenader commented on src/main/scala/com/box/castle/router/kafkadispatcher/cache/CacheWithData.scala:
```diff
@@ -74,43 +75,83 @@ private[cache] class CacheWithData private(val data: LinkedHashMap[Long, CastleM
-          case None => findOption(values.tail)(f)
-        }
-      case None => None
+  private def getMoreRecursive(currentList: Vector[CastleMessageBatch], bufferSize: Int): Vector[CastleMessageBatch] = {
```
this needs protective checking that we are actually returning contiguous messages, if the next batch in line is not contiguous with the current list, we should bail

denisgrenader replied:
You are right, I missed that part.  Can you add some tests to go over the tricky part of not adding a batch if the last batch's size is slightly bigger than the buffer size (and off by 1 errors around that condition checking)

denisgrenader commented on src/test/scala/com/box/castle/batch/CastleMessageBatchTest.scala:
```diff
@@ -49,6 +49,23 @@ class CastleMessageBatchTest extends Specification with Mockito with MockBatchTo
+      combinedMessageBatch.sizeInBytes must_== 100 + 150 + 200 + MessageSet.LogOverhead * 3
+      combinedMessageBatch.offset must_== 50
+      combinedMessageBatch.nextOffset must_== 53
+    }
```
after adding contiguous checking, add tests ensure that when the batches are not contiguous we don't return the vectors that skip messages

shubhrojroy replied:
Done

denisgrenader commented on src/test/scala/com/box/castle/router/kafkadispatcher/cache/FetchDataProcessorCacheTest.scala:
```diff
@@ -138,5 +138,16 @@ class FetchDataProcessorCacheTest extends Specification with Mockito with MockBa
+      val cache = FetchDataProcessorCache(57 + 79 + 219)
+        .add(topicAndPartition, createBatch(20, 2, 57))
+        .add(topicAndPartition, createBatch(22, 1, 79))
+        .add(topicAndPartition, createBatch(23, 7, 219))
```
add checks to make sure the right thing happens when the batches are not contiguous

shubhrojroy replied:
Done

shubhrojroy commented on src/main/scala/com/box/castle/router/kafkadispatcher/cache/CacheWithData.scala:
```diff
@@ -74,43 +75,83 @@ private[cache] class CacheWithData private(val data: LinkedHashMap[Long, CastleM
-          case None => findOption(values.tail)(f)
-        }
-      case None => None
+  private def getMoreRecursive(currentList: Vector[CastleMessageBatch], bufferSize: Int): Vector[CastleMessageBatch] = {
```
Here we querying the cache with the nextOffset of the last batch. This will ensure that the batches are contiguous right ? If there is no batch  in the cache that starts with nextOffset of the last batch we return the list upto the last batch. Lemme know if am missing a corner case here.

shubhrojroy commented:
- Added method CastleMessageBatch.createBatchBySize
This method slices a CastleMessageBatch to fit initial part of it that fits in provided batchSize.

- Added logic to CacheWithData.getAll to fit part of a batch in the cache if it does not fit entirely. 

- Added tests for newly added method and additional tests as suggested by Denis

denisgrenader commented:
Probably should have second pair of eyes on this, did @bharathi-ravishanker or @dashk have a look at this review?

imwmwm commented on src/main/scala/com/box/castle/batch/CastleMessageBatch.scala:
```diff
@@ -74,6 +74,30 @@ case class CastleMessageBatch(messageAndOffsetSeq: IndexedSeq[MessageAndOffset],
+    else {
+      val (endIdx, _) = messageAndOffsetSeq.foldLeft((0, batchSize)){ case ((idx, bSize), message) =>
+        val remainingSize = bSize - MessageSet.entrySize(message.message)
+        (if( remainingSize < 0 ) idx  else idx + 1, remainingSize)
```
when remainingSize becomes less than zero, foldLeft will keep folding until end of the collection even though idx will not change, right?

shubhrojroy replied:
Yes.

denisgrenader commented on src/main/scala/com/box/castle/router/kafkadispatcher/cache/CacheWithData.scala:
```diff
@@ -74,43 +75,103 @@ private[cache] class CacheWithData private(val data: LinkedHashMap[Long, CastleM
-        f(h) match {
-          case Some(v) => Some(v)
-          case None => findOption(values.tail)(f)
+  private def getMoreRecursive(currentList: Vector[CastleMessageBatch], bufferSize: Int): Vector[CastleMessageBatch] = {
```
use foldLeft w/ a tuple here

denisgrenader commented on src/main/scala/com/box/castle/consumer/CastleSimpleConsumer.scala:
```diff
@@ -36,6 +36,12 @@ class CastleSimpleConsumer(broker: Broker,
+    * Returns the consumer bufferSize
+    * @return
+    */
+  def getBufferSize = bufferSize
```
just add val to bufferSize in the constructor

denisgrenader commented on src/main/scala/com/box/castle/router/kafkadispatcher/cache/CacheWithData.scala:
```diff
@@ -74,43 +75,103 @@ private[cache] class CacheWithData private(val data: LinkedHashMap[Long, CastleM
+    * @param bufferSize
+    * @return
+    */
+  def getAll(offset: Long, bufferSize: Int): Option[CastleMessageBatch] = {
```
move this into get

denisgrenader commented on src/main/scala/com/box/castle/router/kafkadispatcher/cache/CacheWithData.scala:
```diff
@@ -5,10 +5,11 @@ import com.box.castle.collections.immutable.LinkedHashMap
- */
+  * An immutable implementation of Cache with data in it.
+  */
 private[cache] class CacheWithData private(val data: LinkedHashMap[Long, CastleMessageBatch],
```
add bufferSize to this constructor, and go up the stack for configuration

denisgrenader commented on src/main/scala/com/box/castle/batch/CastleMessageBatch.scala:
```diff
@@ -74,6 +74,30 @@ case class CastleMessageBatch(messageAndOffsetSeq: IndexedSeq[MessageAndOffset],
+    else {
+      val (endIdx, _) = messageAndOffsetSeq.foldLeft((0, batchSize)){ case ((idx, bSize), message) =>
+        val remainingSize = bSize - MessageSet.entrySize(message.message)
+        (if( remainingSize < 0 ) idx  else idx + 1, remainingSize)
```
separate this into two lines

denisgrenader replied:
```
    val s: IndexedSeq[MyClass] = List(MyClass("bob", 3), MyClass("Shubhro", 7), MyClass("Denis", 5)).toIndexedSeq

    var remains = 9
    val newOne = s.takeWhile(m => {
      remains -= m.size
      remains >= 0
    })
```

shubhrojroy commented on src/main/scala/com/box/castle/router/kafkadispatcher/cache/CacheWithData.scala:
```diff
@@ -43,26 +44,26 @@ private[cache] class CacheWithData private(val data: LinkedHashMap[Long, CastleM
+    * get(26) = None
+    *
+    * @return
+    */
```
Sry I think this got auto-formatted by IntelliJ. Will try to fix it.

shubhrojroy commented:
Addressed code review changes:
- Moved bufferSize to the constructor of the Cache.
- Changed CastleMessageBatch.createBatchBySize to use takeWhile instead of foldLeft.
- Updated existing test due to bufferSize change.

samdehouck commented:
As we discussed, I'm concerned that we will be fragmenting batching for not in-sync committers and that they'll never be in-sync, ie align on the beginning on the batches fetched from Kafka. Is this a performance concern?

shubhrojroy commented:
Discussed this with Denis and we decided to take the batch slicing out for now and observe if its really needed. Academically it maybe the right thing to do, but practically its risky and may not make much of a difference if we took it out esp. given the scenario that this fix is trying to address. Will push the code changes to back the slicing out soon.

samdehouck commented on src/main/scala/com/box/castle/router/kafkadispatcher/cache/CacheWithData.scala:
```diff
@@ -23,112 +25,173 @@ private[cache] class CacheWithData private(val data: LinkedHashMap[Long, CastleM
-        f(h) match {
-          case Some(v) => Some(v)
-          case None => findOption(values.tail)(f)
+  private def getMoreRecursive(currentList: IndexedSeq[CastleMessageBatch],
```
why do you need an IndexedSeq here? It seems a Seq would be enough.

shubhrojroy replied:
In line 136 I am doing an append at the end of currentList which I think is more performant with IndexedSeq than Seq.

samdehouck commented on src/main/scala/com/box/castle/batch/CastleMessageBatch.scala:
```diff
@@ -90,6 +90,22 @@ object CastleMessageBatch {
+    * @param messageBatches
+    * @return
+    */
+  def apply(messageBatches: IndexedSeq[CastleMessageBatch]): CastleMessageBatch = {
```
why not just a Seq?
